### PR TITLE
fix at::linear numeric accuracy for functorch, add test

### LINKED ISSUE #780: Slight numerical differences when using `linear` with functorch transforms
It looks like functorch has its own copy of the C++ decomposition for `linear`, for performance reasons: https://github.com/pytorch/functorch/blob/main/functorch/csrc/PyTorchOperatorHacks.cpp

It looks like the implementations have diverged over time, leading to numeric differences: https://github.com/pytorch/pytorch/blob/56bed0dcfe7ca9047e5c95a6f3d7fcb0ec403b0c/aten/src/ATen/native/Linear.cpp#L37.

You can see an example below, where when I run `linear` with and without `functionalize()`, I get an `addmm` in one case and an `mm` + `add` in another case, leading to a slight numeric difference.

Should there be a better long term plan for keeping these in sync? Maybe a numeric accuracy test for the ops that are copied over?

===== code =====

import torch
from functorch import make_fx
from functorch.experimental import functionalize

def f(x, y, z):
    return torch._C._nn.linear(x, y, z)

x = torch.randn(14, 1, 384)
y = torch.randn(96, 384)
z = torch.randn(96)

out = make_fx(f)(x, y, z)
print(out.graph)

out = make_fx(functionalize(f))(x, y, z)
print(out.graph)

a = f(x, y, z)
b = functionalize(f)(x, y, z)
print(f"max difference: {torch.max(torch.abs(a - b))}")
print(torch.allclose(a, b))

==== outputs ====

graph():
    %x_1 : [#users=1] = placeholder[target=x_1]
    %y_1 : [#users=1] = placeholder[target=y_1]
    %z_1 : [#users=1] = placeholder[target=z_1]
    %t : [#users=1] = call_function[target=torch.ops.aten.t](args = (%y_1,), kwargs = {})
    %view : [#users=1] = call_function[target=torch.ops.aten.view](args = (%x_1, [14, 384]), kwargs = {})
    %addmm : [#users=1] = call_function[target=torch.ops.aten.addmm](args = (%z_1, %view, %t), kwargs = {})
    %view_1 : [#users=1] = call_function[target=torch.ops.aten.view](args = (%addmm, [14, 1, 96]), kwargs = {})
    return view_1
graph():
    %x_1 : [#users=1] = placeholder[target=x_1]
    %y_1 : [#users=1] = placeholder[target=y_1]
    %z_1 : [#users=1] = placeholder[target=z_1]
    %t : [#users=1] = call_function[target=torch.ops.aten.t](args = (%y_1,), kwargs = {})
    %view : [#users=1] = call_function[target=torch.ops.aten.view](args = (%x_1, [14, 384]), kwargs = {})
    %mm : [#users=1] = call_function[target=torch.ops.aten.mm](args = (%view, %t), kwargs = {})
    %_unsafe_view : [#users=1] = call_function[target=torch.ops.aten._unsafe_view](args = (%mm, [14, 1, 96]), kwargs = {})
    %add : [#users=1] = call_function[target=torch.ops.aten.add](args = (%_unsafe_view, %z_1), kwargs = {})
    return add
max difference: 3.814697265625e-06
False

### PR DESCRIPTION
**Pull request by bdhirsh**
This should fix #780.

cc @Chillee @zou3519, would it be reasonable to make the `aten::linear` op composite compliant instead, so you don't have to worry about keeping the code in sync from in core? Not 100% sure but I think you can do the same "if isSubclassLike()" stuff in `linear`.

### CODE REVIEW

**Comment by Chillee** [comment]
Yeah I think it would. But LGTM!

**Comment by zou3519** [comment]
> would it be reasonable to make the aten::linear op composite compliant instead, so you don't have to worry about keeping the code in sync from in core? Not 100% sure but I think you can do the same "if isSubclassLike()" stuff in linear.

Yes
